modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpRequestHandler.java

```diff
@@ -23,23 +23,40 @@
 import io.netty.channel.ChannelHandlerContext;
 import io.netty.channel.SimpleChannelInboundHandler;
 import org.elasticsearch.ExceptionsHelper;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.network.CloseableChannel;
+import org.elasticsearch.http.CorsHandler;
+import org.elasticsearch.http.HttpChannel;
 import org.elasticsearch.http.HttpPipelinedRequest;
+import org.elasticsearch.http.HttpRequest;
+import org.elasticsearch.http.HttpResponse;
+import org.elasticsearch.http.HttpUtils;
 
 @ChannelHandler.Sharable
 class Netty4HttpRequestHandler extends SimpleChannelInboundHandler<HttpPipelinedRequest> {
 
+    private static final ActionListener<Void> NO_OP = ActionListener.wrap(() -> {});
+
     private final Netty4HttpServerTransport serverTransport;
+    private final CorsHandler corsHandler;
 
-    Netty4HttpRequestHandler(Netty4HttpServerTransport serverTransport) {
+    Netty4HttpRequestHandler(Netty4HttpServerTransport serverTransport, CorsHandler corsHandler) {
         this.serverTransport = serverTransport;
+        this.corsHandler = corsHandler;
     }
 
     @Override
     protected void channelRead0(ChannelHandlerContext ctx, HttpPipelinedRequest httpRequest) {
         final Netty4HttpChannel channel = ctx.channel().attr(Netty4HttpServerTransport.HTTP_CHANNEL_KEY).get();
         boolean success = false;
         try {
-            serverTransport.incomingRequest(httpRequest, channel);
+            HttpResponse earlyResponse = corsHandler.handleInbound(httpRequest);
```

Comment:
Maybe we could just move this whole logic into `org.elasticsearch.http.AbstractHttpServerTransport#incomingRequest` to make things even drier?